### PR TITLE
Admin interface for managing A/B tests

### PR DESCRIPTION
**Pull request by bistline**
#### BACKGROUND
#1721 introduced a simple framework for running A/B tests.  However, that framework is limited to only 2 distinct groups (`control` and `intervention`), and enabling/disabling A/B tests can only be done through the Rails console.  Additionally, getting counts of users in each group could only be done through manual database queries or externally via Mixpanel reports.

#### CHANGES
This update adds the `AbTest` class to manage global configurations for individual A/B test instances, as well as an admin interface for configuring tests.  Admins can perform the following functions through a browser:

- Create new A/B test instances
- Enable/disable A/B test instances
- Configure group names, including adding addition groups
- See counts of user assignments
- Add their user account to specific groups to test UX changes
- Remove A/B test instances

When configuring group names:
- If an existing group is renamed, all users assigned to that group will have their assignment updated to reflect the new name
- If an existing group is removed, all users assigned to that group are randomly reassigned to new groups the next time they visit SCP

Short video highlighting main features:

https://user-images.githubusercontent.com/729968/218846293-418b2756-2282-4d8c-9025-c1d87aa2cba7.mov

#### MANUAL TESTING
1. Boot as normal and sign in
2. Go to the "Feature Flags" admin page from the profile menu
3. For any existing feature flag, click the gear icon under the "A/B Test" column
4. Click the "Create A/B Test" button
5. Once created, set the A/B test to be enabled to assign yourself to a given group
6. Use the "Group testing" tab to assign yourself to a different group and note that the user counts changed
7. Add a new group name and confirm you can assign yourself to that
8. Delete the A/B test to clean up all records

### CODE REVIEW

File: app/views/ab_tests/_ab_test_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    <li class="nav-item active"><a href="#ab-test-options" data-toggle="tab">Options</a></li>
+    <li class="nav-item"><a href="#ab-test-groups" data-toggle="tab">Group testing</a></li>
+  </ul>
+
+  <div class="tab-content">
+    <div class="tab-pane active in top-pad" id="ab-test-options">
+      <%= form_with(model: ab_test, url: update_feature_flag_ab_test_path, id: 'ab-test', class: 'form') do |form| %>
+        <% if ab_test.errors.any? %>
+          <div class="bs-callout bs-callout-danger" id="ab-test-errors-block">
+            <h2><%= pluralize(ab_test.errors.count, "error") %> prohibited this A/B test from being saved:</h2>
```
Comment (eweitz) [inline app/views/ab_tests/_ab_test_form.html.erb:12]: I don't think I know of Rails / Ruby `pluralize`!  Convenient.

File: test/models/ab_test_test.rb [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    @ab_test.reload
+    @user.feature_flag_options.destroy_all
+  end
+
+  after(:all) do
+    @feature_flag.destroy
+  end
+
+  # we can't prove true random assignment, only that over a reasonable amount of iterations we would see a roughly
+  # equal distribution (±2%); for reference, SCP had ~2.5K unique Study overview users/week in 2022
```
Comment (eweitz) [inline test/models/ab_test_test.rb:25]: I can't help myself from linking to this statistic: https://mixpanel.com/s/1hzUx.

File: app/models/ab_test.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    return [] if metrics_uuid.nil?
+
+    where(enabled: true).map { |ab_test| ab_test.assignment(metrics_uuid) }.reject(&:flag_override?).map(&:tag)
+  end
+
+  private
+
+  # ensure consistent group name formatting - all lowercase word characters (plus dashes)
+  def sanitize_group_names
+    group_names.map! { |name| name.gsub(NAME_SANITIZER, '').downcase }.reject!(&:blank?)
```
Comment (eweitz) [inline app/models/ab_test.rb:46]: Thanks for including this!

**Review by eweitz — APPROVED** [review]
Code and video look good!  Your [demo yesterday](https://drive.google.com/file/d/19i28XoSvt7d4gHUoKKuQYXigez_0egyE/view?t=28m53s) was also really helpful.

**Review by jlchang — APPROVED** [review]
Functional review - works as described. Looking forward to seeing it in action!

**Pull request merged**